### PR TITLE
[SPARK-32160][CORE][PYSPARK] Add a config to switch allow/disallow to create SparkContext in executors.

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

This is a follow-up of #28986.
This PR adds a config to switch allow/disallow to create `SparkContext` in executors.

- `spark.driver.allowSparkContextInExecutors`

### Why are the changes needed?

Some users or libraries actually create `SparkContext` in executors.
We shouldn't break their workloads.

### Does this PR introduce _any_ user-facing change?

Yes, users will be able to create `SparkContext` in executors with the config enabled.

### How was this patch tested?

More tests are added.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -1908,4 +1908,11 @@ package object config {
       .version("3.1.0")
       .booleanConf
       .createWithDefault(false)
+
+  private[spark] val ALLOW_SPARK_CONTEXT_IN_EXECUTORS =
+    ConfigBuilder("spark.driver.allowSparkContextInExecutors")
+      .doc("If set to true, SparkContext can be created in executors.")
+      .version("3.0.1")
```
Comment (ueshin) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1915]: I'll submit another PR with the opposite default value against `branch-3.0`.

Comment (ueshin) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1915]: Also I'll add a migration guide later.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by ueshin** [comment]
hmm. A test is still failing in https://github.com/apache/spark/pull/29278/checks?check_run_id=928797731:

```
[info] - SPARK-32175: Plugin initialization should start after heartbeater started *** FAILED *** (32 seconds, 952 milliseconds)
[info]   The code passed to failAfter did not complete within 30000000000 nanoseconds. (SparkSubmitSuite.scala:1438)
[info]   org.scalatest.exceptions.TestFailedDueToTimeoutException:
[info]   at java.lang.Thread.getStackTrace(Thread.java:1559)
[info]   at org.scalatest.concurrent.TimeLimits.failAfterImpl(TimeLimits.scala:234)
[info]   at org.scalatest.concurrent.TimeLimits.failAfterImpl$(TimeLimits.scala:233)
[info]   at org.apache.spark.deploy.SparkSubmitSuite$.failAfterImpl(SparkSubmitSuite.scala:1419)
[info]   at org.scalatest.concurrent.TimeLimits.failAfter(TimeLimits.scala:230)
[info]   at org.scalatest.concurrent.TimeLimits.failAfter$(TimeLimits.scala:229)
[info]   at org.apache.spark.deploy.SparkSubmitSuite$.failAfter(SparkSubmitSuite.scala:1419)
[info]   at org.apache.spark.deploy.SparkSubmitSuite$.runSparkSubmit(SparkSubmitSuite.scala:1438)
[info]   at org.apache.spark.executor.ExecutorSuite.$anonfun$new$17(ExecutorSuite.scala:469)
[info]   at org.apache.spark.executor.ExecutorSuite.$anonfun$new$17$adapted(ExecutorSuite.scala:407)
[info]   at org.apache.spark.SparkFunSuite.withTempDir(SparkFunSuite.scala:170)
[info]   at org.apache.spark.executor.ExecutorSuite.$anonfun$new$16(ExecutorSuite.scala:407)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)

...

[info] *** 1 TEST FAILED ***
[error] Failed: Total 2869, Failed 1, Errors 0, Passed 2868, Ignored 8, Canceled 1
[error] Failed tests:
[error] 	org.apache.spark.executor.ExecutorSuite
[error] (core/test:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 1970 s, completed Jul 30, 2020 6:28:55 PM
```

but I can see the failure in other PRs, e.g., https://github.com/apache/spark/pull/29276/checks?check_run_id=924258317.
Also it passes in my local.

I'll rerun the tests.

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by HyukjinKwon** [comment]
GitHub Actiosn build passed. Merged to master.

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @ueshin . Why do we use `spark.driver` namespace in this new configuration? Can we use `spark.executor` namespace instead because this is irrelevant to Spark Driver.

```
spark.driver.allowSparkContextInExecutors
```

Since this is a configuration namespace issue, cc @cloud-fan and @gatorsmile .

I guess we can change like this simply.
```
- spark.driver.allowSparkContextInExecutors
+ spark.executor.allowSparkContext
```

**Comment by ueshin** [comment]
@Dooyoung-Hwang The name was inspired by the old config `spark.driver.allowMultipleContexts` in the comment https://github.com/apache/spark/pull/28986#discussion_r461194397.
But, yeah, `spark.executor.allowSparkContext` sounds more sense. cc @cloud-fan, @gatorsmile

**Comment by ueshin** [comment]
I submitted PRs to rename the config #29340, #29341.

**Comment by dongjoon-hyun** [comment]
Thanks, @ueshin . BTW, my name is `Dongjoon Hyun` with id @dongjoon-hyun . :)

**Comment by ueshin** [comment]
Oops, sorry @dongjoon-hyun and @Dooyoung-Hwang. 🙏

**Comment by HyukjinKwon** [comment]
Actually, what @dongjoon-hyun pointed out makes sense to me. Why don't we just change?

**Comment by HyukjinKwon** [comment]
oh the PRs are already open.

**Pull request closed without merging**